feat(cli): add --delete-namespace to delete namespace incase it is empty after package deletion

Pavan-Chaithanya opened the pull request:
<!-- Thanks for creating this pull request 🤗 Please make sure you followed the conventional commit -->

<!-- If this pull request closes an issue, please mention the issue number below -->
Closes https://github.com/glasskube/glasskube/issues/1102

## 📑 Description
This PR introduces a new flag `--delete-namespace` to the `uninstall` command. When this flag is passed to uninstall a package, the following can happen:
1. If no package is present in the namespace other than the package we are trying to uninstall, It uninstalls the package and deletes the namespace.
2. If other packages are present in the namespace, it returns an error to the CLI. In this case, you might wanna consider deleting the package without this flag.

## ✅ Checks
- [ ] My code requires changes to the documentation
- [ ] I have updated the documentation as required

## ℹ Additional Information
<!-- Any additional information like breaking changes, dependencies added, screenshots, comparisons between new and old behavior, etc. -->

kosmoz commented on cmd/glasskube/cmd/uninstall.go:
```diff
@@ -64,16 +67,39 @@ var uninstallCmd = &cobra.Command{
 					}
 				}
 			}
+		} else if uninstallCmdOptions.DeleteNamespace {
+			// list all packages in the namespace
+			lister := list.NewListerWithRepoCache(ctx)
+			namespace := pkg.GetNamespace()
+			pkgsInNamespace, err := lister.GetPackagesWithStatus(ctx, list.ListOptions{
+				Namespace: namespace,
+			})
+			if err != nil {
+				fmt.Fprintf(os.Stderr, "❌ Could not list packages: %v\n", err)
+				cliutils.ExitWithError()
+			}
+			pkgCount := 0
+			for _, p := range pkgsInNamespace {
+				pkgCount += len(p.Packages)
+			}
+			if pkgCount > 1 {
+				fmt.Fprintf(os.Stderr, "❌ Namespace %v contains more than one package. "+
+					"Current count of packages present in the namespace: %v\n", namespace, pkgCount)
+				cliutils.ExitWithError()
+			}
+			deleteNamespace = true
```
This seems like a lot of unnecessary work. It would be better to use a simple packageClient and just fetch the Package resources directly, rather than using the list package, since that one goes through some extra hoops.

Pavan-Chaithanya replied:
Makes sense. Fixed it

Pavan-Chaithanya commented on pkg/uninstall/uninstall.go:
```diff
@@ -31,23 +32,43 @@ func (obj *uninstaller) WithStatusWriter(sw statuswriter.StatusWriter) *uninstal
+	pkg ctrlpkg.Package,
+	deleteNamespace bool,
+) error {
+	return obj.Uninstall(ctx, pkg)
```
Removed it

Pavan-Chaithanya commented:
@kosmoz Can you please review this PR?

kosmoz commented on pkg/uninstall/uninstall.go:
```diff
@@ -52,6 +54,32 @@ func (obj *uninstaller) Uninstall(ctx context.Context, pkg ctrlpkg.Package, isDr
+	return nil
+}
+
+func (obj *uninstaller) UninstallAndDeleteNamespaceBlocking(
+	ctx context.Context,
+	pkg ctrlpkg.Package,
+	isDryRun bool,
+) error {
+	namespace := pkg.GetNamespace()
+	if err := obj.UninstallBlocking(ctx, pkg, isDryRun); err != nil {
+		return err
+	}
+	clientset := clicontext.KubernetesClientFromContext(ctx)
+
+	return obj.deleteNamespaceBlocking(ctx, clientset, namespace)
+}
```
I think it would be simpler to add a "DeleteNamespace" option to the existing UninstallBlocking function

kosmoz commented on pkg/uninstall/uninstall.go:
```diff
@@ -52,6 +54,32 @@ func (obj *uninstaller) Uninstall(ctx context.Context, pkg ctrlpkg.Package, isDr
+	if err := obj.UninstallBlocking(ctx, pkg, isDryRun); err != nil {
+		return err
+	}
+	clientset := clicontext.KubernetesClientFromContext(ctx)
+
+	return obj.deleteNamespaceBlocking(ctx, clientset, namespace)
```
clientset should be initialized in deleteNamespaceBlocking

kosmoz commented on cmd/glasskube/cmd/uninstall.go:
```diff
@@ -78,11 +79,26 @@ var uninstallCmd = &cobra.Command{
-				cliutils.ExitWithError()
+			// delete the namespace only if it is namespace scoped and there are no other packages in the namespace since
+			// cluster packages delete the namespace once they are deleted
+			deleteNamespace := uninstallCmdOptions.DeleteNamespace && pkg.IsNamespaceScoped()
```
If DeleteNamespace is set and the package is not namespace-scoped, the command shoul exit with an error

kosmoz commented on cmd/glasskube/cmd/uninstall.go:
```diff
@@ -78,11 +79,26 @@ var uninstallCmd = &cobra.Command{
+			// cluster packages delete the namespace once they are deleted
+			deleteNamespace := uninstallCmdOptions.DeleteNamespace && pkg.IsNamespaceScoped()
+			if deleteNamespace {
+				if err := uninstaller.IsNamespaceSafeToDelete(ctx, pkg); err != nil {
```
I think it might be possible to perform this check in the UninstallBlocking method

kosmoz commented:
Hey @Pavan-Chaithanya, how are you doing? Do you need any help with this PR?

Pavan-Chaithanya commented:
Hey @kosmoz, I'm extremely sorry about the delay. I got occupied with other things. I addressed all the comments. Can you have a look? Btw I made some changes in makefile, Reason:
Previously, we ran the lint with whatever version was installed in bin/. But if one of your team upgrades the version of lint, the user cannot automatically upgrade unless he sees it and manually deletes bin/ and installs it. With this change, we check the version and see if a new version needs to be installed and install it.

Pavan-Chaithanya commented:
> Hey @kosmoz, I'm extremely sorry about the delay. I got occupied with other things. I addressed all the comments. Can you have a look? Btw I made some changes in makefile, Reason: Previously, we ran the lint with whatever version was installed in bin/. But if one of your team upgrades the version of lint, the user cannot automatically upgrade unless he sees it and manually deletes bin/ and installs it. With this change, we check the version and see if a new version needs to be installed and install it.

I fixed it because I faced the issue of having an older version and seeing errors which I didn't see on GitHub actions 😓 . You might wanna consider doing the same for the other version if you think this is ok.

Pavan-Chaithanya commented:
@kosmoz Any updates on this PR?
